pipeline/remove_subscribers.py

```diff
@@ -0,0 +1,88 @@
+"""Removing products data that no longer have active subscriptions in the RDS database."""
+import logging
+from psycopg2.extensions import connection
+from dotenv import load_dotenv
+from connect_to_database import configure_logging, get_connection
+
+
+def get_active_subscriptions(conn: connection) -> set:
+    """Returns all product_ids that have users subscribed to it from subscription schema."""
+    try:
+        with conn.cursor() as cur:
+            cur.execute("""SELECT DISTINCT product_id FROM subscription;""")
+            active_subscriptions = cur.fetchall()
+        return {row[0] for row in active_subscriptions}
+    except Exception as e:
+        logging.error("Error fetching active subscriptions: %s", e)
+        return set()
+
+
+def get_all_products(conn: connection) -> set:
+    """Returns all product_ids from product table."""
+    try:
+        with conn.cursor() as cur:
+            cur.execute("SELECT product_id FROM product;")
+            return {row[0] for row in cur.fetchall()}
+    except Exception as e:
+        logging.error("Error fetching all products: %s", e)
+        return set()
+
+
+def delete_price_changes(cur, unsubscribed_product_ids: list) -> None:
+    """Deletes price changes for unsubscribed products."""
+    query = """DELETE FROM price_changes WHERE product_id = ANY(%s);"""
+
+    cur.execute(query, (unsubscribed_product_ids,))
+    logging.info("Removed price changes for products: %s",
+                 unsubscribed_product_ids)
+
+
+def delete_products(cur, unsubscribed_product_ids: list) -> None:
+    """Deletes unsubscribed products from product table."""
+    query = """DELETE FROM product WHERE product_id = ANY(%s);"""
+    cur.execute(query, (unsubscribed_product_ids,))
+    logging.info("Removed unsubscribed products: %s", unsubscribed_product_ids)
+
+
+def clean_websites(cur) -> None:
+    """Removes websites not referenced in product table."""
+    query = """DELETE FROM website
+    WHERE website_id NOT IN (
+        SELECT DISTINCT website_id
+        FROM product);"""
+    cur.execute(query)
+    logging.info("Unused websites have been removed.")
```

Comment:
These three functions, they're primarily just cursor.execute(). Since they're carrying out the same purpose it could be counted as "repeated code".

Reply:
May also want to consider the same for the "get_all_products" and "get_active_subscriptions"

Reply:
This was fixed in the latest push to the branch.